### PR TITLE
Use geometry coordinate element for meshes

### PR DESCRIPTION
**Pull request by garth-wells**
Consistently requires a 'coordinate element' to construct `mesh::Geometry`.

### CODE REVIEW

File: cpp/dolfinx/fem/ElementDofLayout.cpp [hunk 1]
```diff
@@ -240,40 +240,40 @@ int ElementDofLayout::degree() const
   throw std::runtime_error("Cannot determine degree");
 }
 //-----------------------------------------------------------------------------
-ElementDofLayout fem::geometry_layout(mesh::CellType cell, int num_nodes)
-{
-  // TODO: Fix for degree > 2
+// ElementDofLayout fem::geometry_layout(mesh::CellType cell, int num_nodes)
```
Comment (michalhabera) [inline cpp/dolfinx/fem/ElementDofLayout.cpp:243]: Leftover commented code.

File: cpp/dolfinx/fem/ElementDofLayout.h [hunk 2]
```diff
@@ -193,7 +193,7 @@ class ElementDofLayout
 /// @todo Use UFC coordinate dofmap instead?
 ///
 /// Create ElementDofLayout for scalar Lagrange elements. Use for meshes.
-ElementDofLayout geometry_layout(mesh::CellType cell, int num_nodes);
+// ElementDofLayout geometry_layout(mesh::CellType cell, int num_nodes);
```
Comment (michalhabera) [inline cpp/dolfinx/fem/ElementDofLayout.h:196]: Same leftover here.

File: cpp/dolfinx/fem/utils.cpp [hunk 3]
```diff
@@ -683,15 +677,31 @@ fem::create_coordinate_map(const ufc_coordinate_mapping& ufc_cmap)
          {quadrilateral, mesh::CellType::quadrilateral},
          {hexahedron, mesh::CellType::hexahedron}};
 
+  // Map cell tyee
```
Comment (michalhabera) [inline cpp/dolfinx/fem/utils.cpp:680]: Typo

File: cpp/dolfinx/mesh/Geometry.h [hunk 4]
```diff
@@ -117,14 +119,12 @@ class Geometry
   std::shared_ptr<const common::IndexMap> _index_map;
 
   // The dof layout on the cell
-  fem::ElementDofLayout _layout;
+  // fem::ElementDofLayout _layout;
```
Comment (michalhabera) [inline cpp/dolfinx/mesh/Geometry.h:122]: Leftover

File: python/dolfinx/io.py [hunk 5]
```diff
@@ -40,6 +41,18 @@ def write_function(self, u, t=0.0, mesh_xpath="/Xdmf/Domain/Grid[@GridType='Unif
         super().write_function(u_cpp, t, mesh_xpath)
 
     def read_mesh(self, name="mesh", xpath="/Xdmf/Domain"):
-        mesh = super().read_mesh(name, xpath)
-        mesh.geometry.coord_mapping = fem.create_coordinate_map(mesh)
+        # Read mesh data from file
+        cell_type, x, cells = super().read_mesh_data(name, xpath)
+
+        # Construct the geometry map
```
Comment (michalhabera) [inline python/dolfinx/io.py:47]: Could you use here Mesh() method defined below in dolfinx/mesh.py?

File: python/dolfinx/wrappers/generation.cpp [hunk 6]
```diff
@@ -7,12 +7,13 @@
...
 #include "caster_mpi.h"
 #include <array>
+#include <dolfinx/fem/CoordinateElement.h>
 #include <dolfinx/generation/BoxMesh.h>
 #include <dolfinx/generation/IntervalMesh.h>
 #include <dolfinx/generation/RectangleMesh.h>
 #include <dolfinx/generation/UnitDiscMesh.h>
-#include <dolfinx/mesh/Geometry.h>
-#include <dolfinx/mesh/Topology.h>
+// #include <dolfinx/mesh/Geometry.h>
```
Comment (michalhabera) [inline python/dolfinx/wrappers/generation.cpp:15]: Leftovers

File: python/dolfinx/mesh.py [hunk 7]
```diff
@@ -48,6 +48,27 @@ def locate_entities_geometrical(mesh: cpp.mesh.Mesh,
...
+    """Refine a mesh"""
+    if cell_markers is None:
+        mesh_refined = cpp.refinement.refine(mesh, redistribute)
+    else:
+        mesh_refined = cpp.refinement.refine(mesh, cell_markers, redistribute)
+    mesh_refined._ufl_domain = mesh._ufl_domain
+    return mesh_refined
+
+
+def Mesh(comm, cell_type, x, cells, ghosts, degree=1, ghost_mode=cpp.mesh.GhostMode.none):
```
Comment (michalhabera) [inline python/dolfinx/mesh.py:61]: Add docstring

File: python/test/unit/mesh/test_mesh.py [hunk 8]
```diff
@@ -140,7 +140,7 @@ def _check_ufl_domain(mesh):
         assert mesh.geometry.dim == domain.geometric_dimension()
         assert mesh.topology.dim == domain.topological_dimension()
         assert mesh.ufl_cell() == domain.ufl_cell()
-        assert mesh.id == domain.ufl_id()
+        # assert mesh.id == domain.ufl_id()
```
Comment (michalhabera) [inline python/test/unit/mesh/test_mesh.py:143]: Leftover

File: python/dolfinx/io.py (hunk 5 above)
Comment (garth-wells) [reply python/dolfinx/io.py:47]: Not yet - the method in `dolfinx/mesh.py` wraps the deprecated Mesh constructor in the C++ interface.

**Pull request merged**